### ISSUE TITLE
Required property for ASDF writer library

### ISSUE THREAD

**Issue by embray** [spacetelescope/asdf-standard#94]
Directly related to #3, I think somewhere in the top level of the tree we should *require* a property indicating the software (name and version, at least) that wrote out that file.  Requiring this I think would at least discourage ad-hoc writing of files.  I think a timestamp should go along with this.  This wouldn't help much if users then edit a file by hand, but at least then the timestamp will be different from the file's last modified time (unless the user manually edits the timestamp too, but even then it will probably be wrong).

Maybe if we *really* wanted to be strict we could require that every ASDF writer have a unique signature it uses to sign the file.  This doesn't need to be cryptographically secure or anything--just something good enough to confirm that the file was indeed written by that software (or some knucklehead manually computing the signature using some library's key and updating it by hand).

The point is to have some idea who to blame if invalid ASDF files are being written :)

**Comment by mdboom** [comment]
I like the idea of the required property with the software and version (and maybe a URL to the homepage of the software).  That would certainly have been helpful in VOTable land (where an XML comment with the software name is *encouraged* but not *enforced*).

I'm a little more concerned about a timestamp.  It makes in-place updating of the data blocks more difficult (or at least means that the timestamp would only be good enough to indicate file creation not modification -- or maybe it only indicates modification of the tree).  Or the timestamp could be outside of the tree and of fixed length (maybe right after the `#ASDF` header), in which case it could be edited in-place.

Even a checksum of the YAML content would be good enough to detect if the file was edited by hand (again, assuming someone doesn't go through the trouble of computing it by hand also).  That would allow a library to throw up a big fat warning that "all bets are off", but then still try its best anyway.

And, if I'm following that this idea also somewhat follows from the problems outlined in astropy/astropy#3941, one solution is also to establish a "convention" early on that code to workaround invalid files will not be written.  (Of course, in the real world, I would guess that the spec is probably underspecified in a lot of ways and some things will be invalid "by accident", and we'll have to make clarifications as we go on.  In which case, we're still better off then FITS because we have a version number and can deprecate things).

**Comment by embray** [comment]
My idea about adding a signature using a per-library random key was that the signature would even be checked (in most cases at least) when reading the file.  It's more for debugging purposes--to determine "Was this file really written by that software?"  It's true though that could be a problem for things like in-place updates, and especially if the file is being updated by different software than the one that wrote it.  Need to think about that more...

**Comment by embray** [comment]
> Of course, in the real world, I would guess that the spec is probably underspecified in a lot of ways and some things will be invalid "by accident", and we'll have to make clarifications as we go on. In which case, we're still better off then FITS because we have a version number and can deprecate things

Right.  We can't blame writers for doing bad things because the spec was underspecified.  We can however give them a path to improvement (including fixing up existing files).
